recipes-core/tdx-enc-handler/tdx-enc-handler/tdx-enc.sh

```diff
@@ -37,39 +49,74 @@ tdx_enc_exit_error() {
     exit 1
 }
 
-# system checks
-tdx_enc_check() {
+# All backends: prepare and check system
+tdx_enc_prepare_generic() {
+    tdx_enc_log "Preparing and checking system (generic)..."
+
     if ! dmsetup targets | grep crypt -q; then
         tdx_enc_exit_error "No support for dm-crypt target!"
     fi
+}
+
+# CLEARTEXT: prepare system
+tdx_enc_prepare_cleartext() {
+    tdx_enc_log "Preparing and checking system (cleartext)..."
+}
+
+# CAAM: prepare system
+tdx_enc_prepare_caam() {
+    tdx_enc_log "Preparing and checking system (caam)..."
+
     if ! grep -q cbc-aes-caam /proc/crypto; then
         tdx_enc_exit_error "No support for cbc-aes-caam!"
     fi
 }
 
-# check if the encrypted key exists and create one if needed
-tdx_enc_key_gen() {
-    tdx_enc_log "Checking for the encrypted key..."
+# configure key in kernel keyring
+tdx_enc_keyring_configure() {
+    TDX_ENC_KEY_KEYRING_TYPE="$1"
+    KEYNAME="$2"
+    NEW_KEY_CMD="$3"
+    LOAD_KEY_CMD="$4"
 
-    ENC_KEY_FILE="${TDX_ENC_KEY_DIR}/${TDX_ENC_KEY_FILE}"
+    tdx_enc_log "Configuring key in kernel keyring (type=$TDX_ENC_KEY_KEYRING_TYPE keyname=$KEYNAME)..."
 
-    if [ ! -e "${ENC_KEY_FILE}" ]; then
-        tdx_enc_log "Encrypted key not found. Creating it..."
-        KEY="$(keyctl add trusted tdxenc 'new 32' @s)"
+    if [ ! -e "${TDX_ENC_KEY_FULLPATH}" ]; then
+        tdx_enc_log "Key blob not found. Creating it..."
+        KEYHANDLE="$(keyctl add "${TDX_ENC_KEY_KEYRING_TYPE}" "${KEYNAME}" "$(eval echo ${NEW_KEY_CMD})" @s)"
```

Comment:
Yes, the `eval echo` is really required here.

If the variable would have just a string, then it would not be required, but sometimes the variable has a command that needs to be executed before being passed to the `keyctl` command.

For example, the `tdx_enc_key_gen_tpm()` function will pass "`load $(cat ${TDX_ENC_KEY_FULLPATH}`" to this variable, which needs to be expanded inside the tdx_enc_keyring_configure() function.

So the best solution I found for this is expanding the variable via `eval echo`. 

Obs.: I know shellchek complains about it.